### PR TITLE
[4.0] Runtime: Include source location information in log messages about deprecated implicit Objective-C entry points.

### PR DESCRIPTION
**Pull request by jckarter**
Explanation: Make it easier for migration by pinpointing exactly where to insert @objc to keep the entry points in Swift 4 mode.

Scope: Improves the user experience when migrating from Swift 3 to 4.

Issue: rdar://problem/32230003

Risk: Low, straightforward change to existing logging functionality for deprecated objc entry points.

Testing: Swift CI

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@DougGregor How's this look for the 4.0 branch?

File: lib/SILGen/SILGenBridging.cpp [hunk 1]
```diff
@@ -1050,8 +1050,28 @@ void SILGenFunction::emitNativeToForeignThunk(SILDeclRef thunk) {
       if (attr->isSwift3Inferred() &&
           !decl->getAttrs().hasAttribute<DynamicAttr>() &&
           !getASTContext().LangOpts.isSwiftVersion3()) {
-        B.createBuiltin(loc, getASTContext().getIdentifier("swift3ImplicitObjCEntrypoint"),
-            getModule().Types.getEmptyTupleType(), { }, { });
+        
+        // Get the starting source location of the declaration so we can say
+        // exactly where to stick '@objc'.
+        SourceLoc objcInsertionLoc = decl->getStartLoc();
```
Comment (DougGregor) [inline lib/SILGen/SILGenBridging.cpp]: you want `getAttributeInsertionLoc(/*forModifier=*/false)` here, not `getStartLoc()`.

File: lib/SILGen/SILGenBridging.cpp [hunk 2]
```diff
@@ -1050,8 +1050,28 @@ void SILGenFunction::emitNativeToForeignThunk(SILDeclRef thunk) {
...
+        // For a property accessor, direct the user to put @objc on the property
+        // declaration.
+        // For a var, the @objc has to precede the pattern binding.
+        if (auto var = dyn_cast<VarDecl>(decl)) {
+          objcInsertionLoc = var->getParentPatternBinding()->getStartLoc();
```
Comment (DougGregor) [inline lib/SILGen/SILGenBridging.cpp]: Yeah, `getAttributeInsertionLoc` will just DTRT.

**Review by DougGregor — APPROVED** [review]
We can provide slightly better location info, but otherwise LGTM.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@DougGregor How does this look now?

**Comment by DougGregor** [comment]
LGTM!

**Pull request merged**